test/e2e/storage/testsuites/snapshottable.go

```diff
@@ -79,167 +87,330 @@ func (s *snapshottableTestSuite) SkipRedundantSuite(driver TestDriver, pattern t
+		l.pvc, err = l.cs.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Get(context.TODO(), l.pvc.Name, metav1.GetOptions{})
 		framework.ExpectNoError(err)
 
 		// Get the bound PV
-		_, err = cs.CoreV1().PersistentVolumes().Get(context.TODO(), pvc.Spec.VolumeName, metav1.GetOptions{})
+		_, err = l.cs.CoreV1().PersistentVolumes().Get(context.TODO(), l.pvc.Spec.VolumeName, metav1.GetOptions{})
 		framework.ExpectNoError(err)
+	}
+	cleanup := func(l *snapshottableLocal) {
+		// Depending on how far the test executed, cleanup accordingly
+		// Execute in reverse order, similar to defer stack
+		for i := len(l.cleanupSteps) - 1; i >= 0; i-- {
+			l.cleanupSteps[i]()
+		}
 
-		ginkgo.By("creating a SnapshotClass")
-		vsc, err = dc.Resource(SnapshotClassGVR).Create(context.TODO(), vsc, metav1.CreateOptions{})
-		framework.ExpectNoError(err)
-		defer func() {
-			framework.Logf("deleting SnapshotClass %s", vsc.GetName())
-			framework.ExpectNoError(dc.Resource(SnapshotClassGVR).Delete(context.TODO(), vsc.GetName(), metav1.DeleteOptions{}))
-		}()
+		// All tests will require these driver cleanup tests
+		err := tryFunc(l.driverCleanup)
+		framework.ExpectNoError(err, "while cleaning up driver")
+	}
 
-		ginkgo.By("creating a snapshot")
-		snapshot := getSnapshot(pvc.Name, pvc.Namespace, vsc.GetName())
+	ginkgo.BeforeEach(func() {
+		// Check preconditions.
+		framework.ExpectEqual(pattern.SnapshotType, testpatterns.DynamicCreatedSnapshot)
+		dInfo := driver.GetDriverInfo()
+		ok := false
+		sDriver, ok = driver.(SnapshottableTestDriver)
+		if !dInfo.Capabilities[CapSnapshotDataSource] || !ok {
+			e2eskipper.Skipf("Driver %q does not support snapshots - skipping", dInfo.Name)
+		}
+		dDriver, ok = driver.(DynamicPVTestDriver)
+		if !ok {
+			e2eskipper.Skipf("Driver %q does not support dynamic provisioning - skipping", driver.GetDriverInfo().Name)
+		}
+	})
 
-		snapshot, err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Create(context.TODO(), snapshot, metav1.CreateOptions{})
-		framework.ExpectNoError(err)
-		defer func() {
-			framework.Logf("deleting snapshot %q/%q", snapshot.GetNamespace(), snapshot.GetName())
-			// typically this snapshot has already been deleted
-			err = dc.Resource(SnapshotGVR).Namespace(snapshot.GetNamespace()).Delete(context.TODO(), snapshot.GetName(), metav1.DeleteOptions{})
-			if err != nil && !apierrors.IsNotFound(err) {
-				framework.Failf("Error deleting snapshot %q. Error: %v", pvc.Name, err)
-			}
-		}()
-		err = WaitForSnapshotReady(dc, snapshot.GetNamespace(), snapshot.GetName(), framework.Poll, framework.SnapshotCreateTimeout)
-		framework.ExpectNoError(err)
+	// This intentionally comes after checking the preconditions because it
+	// registers its own BeforeEach which creates the namespace. Beware that it
+	// also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
```

Comment:
Shouldn't line 90: f := framework.NewDefaultFramework("snapshotting") be here after line 189? hence the comment of being "intentionally after checking preconditions". Can you please check

Reply:
Fixed the ordering. Moved the BeforeEach to be before the NewFramework call, similar to the tests in provisioning.go